File: day03/day03.py
import re

text = ""
with open("input.txt", "r") as file:
    for line in file:
        text += line


def authenticate(text):
    regex = re.compile(r"mul\(\d{1,3},\d{1,3}\)")
    mul_results = regex.findall(text)

    multiply_numbers = re.compile(r"\d{1,3}")

    def str_int(split):
        extracted = multiply_numbers.findall(split)
        numbers = [int(a) for a in extracted]
        return numbers[0] * numbers[1]

    results = map(str_int, mul_results)
    return sum(list(results))


print("Part 1:", authenticate(text))


def authenticate_2(text):
    regex = re.compile(r"do\(\)|don't\(\)|mul\(\d{1,3},\d{1,3}\)")
    mul_results = regex.findall(text)

    accept = True

    def extract(entry):
        nonlocal accept
        if not accept and entry[0] == "m":
            return 0
        elif not accept and entry == "do()":
            accept = True
            return 0
        elif not accept and entry == "don't()":
            return 0
        elif accept and entry == "don't()":
            accept = False
            return 0
        elif accept and entry == "do()":
            return 0

        reg_split = r"\d{1,3}"
        reg_compile = re.compile(reg_split)
        numbers = reg_compile.findall(entry)

        return int(numbers[0]) * int(numbers[1])

    results = sum(list(map(extract, mul_results)))


    return results


print(authenticate_2(text))

File: day03/day03.js
const fs = require('fs');
const input = String(fs.readFileSync('input.txt'));

const authenticate = function (input) {
  const re = /mul\(\d{1,3},\d{1,3}\)/g;

  const equations = [...input.matchAll(re)].map(function (entry) {
    return entry[0];
  });

  return equations
    .map(function (entry) {
      let reSplit = /\d{1,3}/g;
      const numbers = [...entry.matchAll(reSplit)].map(function (entry) {
        return Number(entry[0]);
      });
      return numbers.reduce((a, b) => a * b, 1);
    })
    .reduce(function (a, b) {
      return a + b;
    });
};

console.log('Part 1:', authenticate(input));

const authenticatePart2 = function (input) {
  const re = /do\(\)|don't\(\)|mul\(\d{1,3},\d{1,3}\)/g;

  const equations = [...input.matchAll(re)].map(function (entry) {
    return entry[0];
  });

  let accept = true;

  return equations
    .map(function (entry) {
      if (!accept && entry[0] === 'm') {
        return 0;
      } else if (!accept && entry === 'do()') {
        accept = true;
        return 0;
      } else if (!accept && entry === "don't()") {
        return 0;
      } else if (accept && entry === "don't()") {
        accept = false;
        return 0;
      } else if (accept && entry == 'do()') {
        return 0;
      }

      let reSplit = /\d{1,3}/g;
      const numbers = [...entry.matchAll(reSplit)].map(function (entry) {
        return Number(entry[0]);
      });
      return numbers.reduce((a, b) => a * b, 1);
    })
    .reduce(function (a, b) {
      return a + b;
    });
};

console.log('Part 2:', authenticatePart2(input));
